### PR TITLE
Implement Tensor struct

### PR DESCRIPTION
**Pull request by siliconlad**

### CODE REVIEW

**Comment by siliconlad** [comment]
@CameronMatthew can you have a look and see if this is heading in the right direction.

Very rough around the edges (can you flatten an n-dimensional tensor in your head?), but that can be patched up with refinements to the API and maybe some additional macros.

Also code is a bit of a mess (wth is `iters.rs`) but gotta start somewhere

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,924 @@
...
+
+    /// For the maths see: https://bit.ly/3KQjPa3
+    fn calculate_index(&self, indices: &[usize]) -> usize {
+        let mut index = 0;
+        for k in 0..self.shape.len() {
+            let stride = self.shape.dims[k+1..].iter().product::<usize>();
+            index += indices[k] * stride;
+        }
+        index
+    }
```
Comment (CameronMatthew) [inline src/lib.rs:239]: This would potentially be very useful to abstract away for other Tensor types

**Review by CameronMatthew — COMMENTED** [review]
Definitely along the right lines! This is very nice. One or two places you can make your life easier. Would also like to talk about reusing some of this for Static tensors

File: src/iter.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
+pub struct IndexIterator {
+    shape: Vec<usize>,
+    current: Vec<usize>,
+    done: bool,
```
Comment (Dougal-s) [inline src/iter.rs:4]: Dynamically allocating memory for every iterator seems like it could be a significant performance problem. I think it would be better to pass the order of the tensor as a template parameter.
```rust
pub struct IndexIterator<const N: usize> {
    shape: [usize; N], // should this be a Shape?
    current: [usize; N],
    done: bool,
}
```
The same thing applies to `Shape` and `Tensor`. Plus, it adds some more type safety.
Granted this does mean that you can't have arrays of different sized tensors, but you can always add extra dimensions of size 1.

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,733 @@
...
+    data: Vec<T>,
+    shape: Shape
+}
+
+impl<T: Num + PartialOrd + Copy> Tensor<T> {
+
+    pub fn new(shape: &Shape, data: &[T]) -> Tensor<T> {
+        assert!(data.len() == shape.size());
+        Tensor {data: data.to_vec(), shape: shape.clone()}
+    }
```
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs:26]: BUT THEY GAVE INCONSISTENT SHAPE INFORMATION AGGGH.

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,789 @@
...
+        let total_size = shape.size();
+        let mut vec = Vec::with_capacity(total_size);
+        for _ in 0..total_size { vec.push(value); }
+        Tensor::new(shape, &vec)
+    }
+
+    pub fn zeros(shape: &Shape) -> Tensor<T> {
+        Tensor::fill(shape, T::zero())
+    }
+
```
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs:38]: Done.

File: src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,789 @@
...
+    pub fn eye(shape: &Shape) -> Tensor<T> {
+        let mut t = Tensor::zeros(shape);
+        for i in 0..shape.dims[0] {
+            t.set_element(&[i, i], T::one());
+        }
+        t
+    }
```
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs:49]: When your understanding isn't good enough you accidentally get it correct :smile:

File: src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,989 @@
...
+        result
+    }
+}
+
+// Vector/Matrix Multiplication
+impl<T: Num + PartialOrd + Copy> Mul<Tensor<T>> for Tensor<T> {
+    type Output = Tensor<T>;
+
+    fn mul(self, rhs: Tensor<T>) -> Tensor<T> {
+    if self.shape.len() == 1 && rhs.shape.len() == 1 {
```
Comment (Dougal-s) [inline src/lib.rs:294]: I think a lot these cases could just be merged? At least, I think that's the case if the code differentiates between row and column vectors.

File: src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,789 @@
...
+        let mut t = Tensor::zeros(shape);
+        for i in 0..shape.dims[0] {
+            t.set_element(&[i, i], T::one());
+        }
+        t
+    }
+
+    // Element-wise operations
+    pub fn pow(&self, power: usize) -> Tensor<T> {
+        let mut t = self.clone();
```
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs:53]: I feel like by default they should create copies. Perhaps a `pow_mut` for in-place ops?

File: src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,789 @@
...
+        t
+    }
+
+    // Properties
+    pub fn shape(&self) -> &Shape { &self.shape }
+    pub fn size(&self) -> usize { self.shape.size() }
+
+    // Access methods
+    pub fn get_element(&self, indices: &[usize]) -> &T {
+        self.assert_indices(indices).unwrap();
```
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs:66]: BUT THEY GAVE THE WRONG INDICES AAGH.

File: src/shape.rs [hunk 9]
```diff
@@ -0,0 +1,74 @@
+use std::fmt;
+use std::ops::Index;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Shape {
```
Comment (Dougal-s) [inline src/shape.rs:5]: Do we want to treat `N x 1` as different from just `N`?

File: src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,924 @@
...
+                    indices.insert(axis, sum_index[i]);
+                }
+
+                let value = *t.get_element(&target) + *self.get_element(&indices);
+                t.set_element(&target, value);
+            }
+        }
+
+        t
+    }
```
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs:106]: I'm just surprised I got it working at all.

File: src/lib.rs (hunk 1 above)
Comment (siliconlad) [reply src/lib.rs:260]: You mean like sparse tensors?

File: src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,924 @@
...
+            for j in 0..self.shape[1] {
+                sum = sum + self.data[i * self.shape[1] + j] * rhs.data[j];
+            }
+            result.data[i] = sum;
+        }
+        result
+    } else if self.shape.len() == 2 && rhs.shape.len() == 2 {
+        // Matrix-Matrix multiplication
+        assert!(self.shape[1] == rhs.shape[0], "The number of columns in the first matrix must be equal to the number of rows in the second matrix.");
+        let mut result = Tensor::zeros(&shape![self.shape[0], rhs.shape[1]]);
```
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs:329]: Broadcasting and Tensor slices would be next on the to do list.

File: src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,989 @@
...
+use std::ops::Mul;
+use std::ops::Div;
+
+use crate::shape::Shape;
+use crate::iter::IndexIterator;
+
+pub type Axes = Vec<usize>;
+
+#[derive(Debug, Clone)]
+pub struct Tensor<T: Num> {
```
Comment (Dougal-s) [inline src/lib.rs:16]: I think it would be best if we had an `AbstractTensor` trait that `Tensor` implements since we'll have various different types of tensors (e.g. `SparseTensor`, `StaticTensor`).

File: src/lib.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (CameronMatthew) [reply src/lib.rs:329]: Speaking of a todo list...

We don't have one.

It will be very easy to get carried away and just implement a bunch of stuff. We should come up with a set of operations to stop at and add more later if there's appetite.

I will start one in an issue.

File: src/lib.rs (hunk 1 above)
Comment (CameronMatthew) [reply src/lib.rs:260]: Sparse and static.

Comment (CameronMatthew) [reply src/lib.rs:260]: Actually mostly for static as sparse tensors are a different beast

File: src/lib.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs:329]: Yeah I guess that's what we can use this initial implementation for.

**Review by Dougal-s — COMMENTED** [review]
I'm very impressed by how quickly you managed to put this together!

File: src/lib.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (CameronMatthew) [reply src/lib.rs:329]: https://github.com/Rust-Scientific-Computing/feotensor/issues/1#issuecomment-2183283636

File: src/lib.rs (hunk 1 above)
Comment (CameronMatthew) [reply src/lib.rs:260]: Actually, this could work well - why don't we make this a method for the Shape struct. Since technically all you need is a shape to work out the right index?

File: src/lib.rs (hunk 6 above)
Comment (siliconlad) [reply src/lib.rs:294]: Yep, definitely needs a rewrite.

File: src/shape.rs (hunk 9 above)
Comment (siliconlad) [reply src/shape.rs:5]: Good question. What do you think?

File: src/lib.rs (hunk 12 above)
Comment (siliconlad) [reply src/lib.rs:16]: Yeah I agree. Was thinking of calling it `RawTensor`. That might be semantically slightly different though.

File: src/lib.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs:106]: Wouldn't this require N more memory allocations. One for each intermediate stage?

[the remark this answers is not in the document]
Comment (CameronMatthew) [reply src/lib.rs:106]: That's true, but is that performance gain worth the sacrifice in complexity/readability at this stage? I'd rather we spent the effort making a sensible abstraction that allowed you to easily do any generic reduction function. Arguably you're optimising for a case that is rare in scientific computing. I make that claim based on very little knowledge of the field in fairness so happy to be wrong here.

[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs:106]: Yeah these methods need to be refactored haha.

File: src/lib.rs (hunk 12 above)
Comment (CameronMatthew) [reply src/lib.rs:16]: That sounds a lot like inheritance - which is not what traits are for. Rather we should have SparseTensors, StaticTensors and DynamicTensors which are composed of reusable parts (like Shape _maybe_).

File: src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs:49]: I'm looking at this again, and how is it right :laughing:

What exactly do we want this method to return?

File: src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs:53]: Let's create copies for now. I think this is the better default.

File: src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,789 @@
...
+        if indices.len() != self.shape.len() {
+            return Err(format!("Incorrect number of dimensions ({} vs {}).", indices.len(), self.shape.len()));
+        }
+        for (i, &index) in indices.iter().enumerate() {
+            if index >= self.shape.dims[i] {
+                return Err(format!("Index out of bounds for dimension {}", i));
+            }
+        }
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs:272]: One advantage of doing it this way is that I can give more helpful errors. Simplifies the `get` and `set` methods as well as I don't need conditional if statements.

File: src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs:49]: Deleted it for now. Will see if I can implement it just for matrices later.

File: src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,924 @@
...
+// Element-wise Multiplication
+impl<T: Num + PartialOrd + Copy> Mul<T> for Tensor<T> {
+    type Output = Tensor<T>;
+
+    fn mul(self, rhs: T) -> Tensor<T> {
+        let mut result = Tensor::zeros(&self.shape);
+        for i in 0..self.size() {
+            result.data[i] = self.data[i].clone() * rhs;
+        }
+        result
```
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs:285]: See above.

File: src/lib.rs (hunk 1 above)
Comment (siliconlad) [reply src/lib.rs:260]: We agreed that this should go in a `RawTensor` struct which is responsible for laying out the tensor in memory.

File: src/shape.rs (hunk 9 above)
Comment (siliconlad) [reply src/shape.rs:5]: I think the consensus was that `N x 1` (and `1 x N`) and `N` are different because they have different "orders". Having said that, I think we will have something which means that `N` can be used in matrix multiplication and other operations as long as it's unambiguous.

For example `N` should be able to multiply with an `N x N` matrix. Whether it's treated as `N x 1` or `1 x N` being based on the order of the multiplication.

The return order is a bit hazy, because numpy would return an order 1 matrix if we use `N`. So maybe we make it consistent with that.

File: src/lib.rs (hunk 12 above)
Comment (siliconlad) [reply src/lib.rs:16]: We probably will want a trait which outlines the basic operations for a generic tensor.

File: src/lib.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs:272]: Refactored this away.

**Comment by siliconlad** [comment]
I have restructured the code to have some of the features that we want. Namely`DynamicStorage` which abstracts away the memory mapping of the tensor. I have also tweaked some of the interfaces for the methods as mentioned above (e.g. returning `Result` in various places).

I'm going to press ahead with some further restructurings:
- [ ] Create dedicated `Vector` and `Matrix` types which support some extra methods
- [ ] Create a distinction between int and float Tensors (to hopefully support `pow)

I'm ignoring the specific implementations of the methods for now because they work (even if they are over-complicated). I will revisit them at the end and tidy them up as well as implement any of the other missing functionality.

**Comment by siliconlad** [comment]
Would be good to have another review of the code written so far. I've implemented most of the methods outlined in #1 so far (missing more of the advanced features). Before I add any more, would like to get some feedback on the structure of everything. The way I've done it seems to work, but there's a fair bit of duplication. Perhaps there's a better way.

Largely ignore the code within the methods for now. A lot of it needs to be looked at and refactored. They do get the job done though so :shrug: (I've never had more tests than I have for this repository I don't think ahaha).

@CameronMatthew @Dougal-s

File: src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,789 @@
...
+    pub fn eye(shape: &Shape) -> Tensor<T> {
+        let mut t = Tensor::zeros(shape);
+        for i in 0..shape.dims[0] {
+            t.set_element(&[i, i], T::one());
+        }
+        t
+    }
+
+    // Element-wise operations
+    pub fn pow(&self, power: usize) -> Tensor<T> {
```
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs]: A better `pow` has been implemented, but it is only available if your data type implements the `num::Float` trait, which I think is reasonable for now.

File: src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,924 @@
...
+use num::Num;
+use std::ops::Add;
+use std::ops::Sub;
+use std::ops::Mul;
+use std::ops::Div;
+
+use crate::shape::Shape;
+use crate::iter::IndexIterator;
+
+pub type Axes = Vec<usize>;
```
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs]: We will still need a vector of axes even if we go with repeated single-axis contractions, unless you want the API to only support one axis at a time (which seems annoying to me). At that point, idk if you want an explicit type like this, or 'just' a `Vec<usize>`.

File: src/lib.rs (hunk 6 above)
Comment (siliconlad) [reply src/lib.rs]: They've been split across the different methods in `DynamicVector`, `DynamicMatrix` and `DynamicTensor`.

**Comment by CameronMatthew** [comment]
I had a quick look and seems fine. I think we should just merge this asap now.

For the following reasons...
- Something "imperfect" is much better than nothing.
- This PR is getting too large to review properly.
- While this remains on a dev branch, the work you do on it is so broadly scoped that it is impossible for others to contribute without risking massive merge conflicts.
- Merging will allow us to create small, self-contained issues that we can all pick up, work on and review easily.

TL;DR Ship it 🚢.

**Comment by siliconlad** [comment]
Sounds good to me.

**Review by CameronMatthew — APPROVED** [review]
LGTM!

File: src/lib.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs]: See #11

[the remark this answers is not in the document]
Comment (siliconlad) [reply src/lib.rs]: See #11
